File: kpca_function.py

#Author: Pablo Campos V.

from scipy.spatial.distance import pdist, squareform
from scipy import exp
from scipy.linalg import eigh
import numpy as np
import matplotlib.pyplot as plt
from scipy.sparse.linalg import eigsh


def stepwise_kpca(X, gamma, n_components):
    """
    Implementation of a RBF kernel PCA.

    Arguments:
        X: A MxN dataset as NumPy array
        gamma: Parameter for the RBF kernel.
        n_components: The number of components to be returned.

    Returns the k eigenvectors (alphas) that correspond to the k largest 
        eigenvalues (lambdas).

    """
    # Squared Euclidean distances for every pair of points
    # in the MxN dimensional dataset.
    sq_dists = pdist(X, 'sqeuclidean')

    # Converting the pairwise distances into a symmetric MxM matrix.
    mat_sq_dists = squareform(sq_dists)

    # Computing the MxM kernel matrix.
    K = exp(-gamma * mat_sq_dists)

    # Centering the kernel matrix.
    N = K.shape[0]
    one_n = np.ones((N,N)) / N
    K_norm = K - one_n.dot(K) - K.dot(one_n) + one_n.dot(K).dot(one_n)

    # Obtaining eigenvalues in descending order with corresponding 
    # eigenvectors from the centered Kernel matrix.
    # Here we use eigsh, which is a wrapper to ARPACK solver for large scale eigenvalue problems.
    eigvals, eigvecs = eigsh(K_norm, k = n_components, which = 'LM')
    
    # Obtaining the i eigenvectors  that corresponds to the i highest eigenvalues (lambdas).
    alphas = np.column_stack((eigvecs[:,-i] for i in range(1,n_components+1)))
    lambdas = [eigvals[-i] for i in range(1,n_components+1)]

    return alphas, lambdas
    
    
#Example: Iris dataset 
from sklearn import datasets

iris = datasets.load_iris()
X = iris.data[:, :4]  
y = iris.target
 
 
gamma = .2
alphas, lambdas = stepwise_kpca(X, gamma, n_components=4)
    

def project_x(x_new, X, gamma, alphas, lambdas):
    pair_dist = np.array([np.sum((x_new-row)**2) for row in X])
    k = np.exp(-gamma * pair_dist)
    return k.dot(alphas / lambdas)
    
def itera(base_new, X, gamma, alphas, lambdas):
     projection = np.array([project_x(row, X, gamma, alphas=alphas, lambdas=lambdas) for row in base_new])
     return projection
   
def grafica(base_projected, i, j):
    plt.plot(base_projected[:, i-1], base_projected[:, j-1], "bo")
    plt.title("Projections")
    plt.xlabel("Component $i$")
    plt.ylabel("Component $j$")
    plt.show()
 
#Call to -itera- function  
base_new_projected = itera(X,X,gamma,alphas,lambdas)   

#Visualization 
grafica(base_new_projected,1,2) 
grafica(base_new_projected,1,3) 
grafica(base_new_projected,1,4) 
grafica(base_new_projected,2,3) 
grafica(base_new_projected,2,4)
grafica(base_new_projected,3,4)

File: kpca_function.R


#Author: Pablo Campos V.

library(kernlab)
library(rARPACK)
library(ggplot2)
library(rgl)


#Let's create some fake data
set.seed(123)
x<- rbind(matrix(rnorm(1200), , 2), matrix(rnorm(1200,mean = 3), , 2))
class <- matrix(c(rep(0, 1000), rep(1, 200)))
df<-as.data.frame(cbind(x,class))
df$class<-df$V3
df$V3<-NULL


#Function 
kpca.hand <- function(mat,k,gamma){
  
  #Implementation of a RBF kernel PCA.
  
  #Arguments:
  #  mat: A MxN dataset as matrix class
  # gamma: Parameter for the RBF kernel.
  # k: The number of components to be returned.
  
  #Returns the k eigenvectors (alphas) that correspond to the k largest 
  #eigenvalues (lambdas)
  
  rbf <- rbfdot(sigma = gamma)
  # Computing the MxM kernel matrix.
  kmat<-kernelMatrix(rbf, mat)
  
  # Centering the kernel matrix.
  ones <- matrix(1/dim(kmat)[1],dim(kmat)[1],dim(kmat)[1])
  kmatcent <- kmat - ones%*%kmat - kmat%*%ones + ones%*%kmat%*%ones
  
  # Obtaining eigenvalues in descending order with corresponding 
  # eigenvectors from the centered Kernel matrix.
  # Here we use eigs, which is a wrapper to ARPACK solver for large scale eigenvalue problems.
  eig <- eigs(kmatcent,k)
  
  return(eig)
}

#Example: Fake data  
gamma <- 0.5
kpca.df <- kpca.hand(as.matrix(df),5,gamma)
#Obtaining the i eigenvectors  that corresponds to the k highest eigenvalues (lambdas)
alphas <- kpca.df$vectors
lambdas <- kpca.df$values

#Plot the first two eigenvectors
plot(kpca.df$vectors[,1],kpca.df$vectors[,2])

qplot(kpca.df$vectors[,1],kpca.df$vectors[,2],color=factor(df$class))
supp.df <- as.data.frame(supp.sc)



project <- function(x_new, Xm, gamma, alphas, lambdas){
  #Function to project new data
  dist2 <- function(y){dist(rbind(x_new,y))^2}
  pair.dist <- apply(Xm,1,dist2)
  k.exp <- exp(-gamma*pair.dist)
  
  alphas.norm <- t(alphas)/lambdas
  alphas.norm <- t(alphas.norm)
  
  aux <- k.exp%*%alphas.norm
  return(aux)
}


#Let's create some fake data for test
set.seed(456)
x_n<- rbind(matrix(rnorm(1200), , 2), matrix(rnorm(1200,mean = 3), , 2))
class_n <- matrix(c(rep(0, 1000), rep(1, 200)))
df_n<-as.data.frame(cbind(x_n,class_n))

#Let's call the -project- function over each test data point
proj.iter.test <- apply(as.matrix(df_n),1,project,Xm=as.matrix(df),gamma=gamma,alphas=alphas,
                        lambdas=lambdas)

test.proj <- t(proj.iter.test)

#Plot the projections into the first two components
qplot(test.proj[,1],test.proj[,2],col=as.factor(df_n$V3),
      xlab="1st Principal Component",ylab="2nd Principal Component")+
  geom_point(size=3)

#Plot the projections into the first three components
plot3d(test.proj[,1],test.proj[,2],test.proj[,3],
       size=3,col=as.factor(df_n$V3+1))